types/wlr_virtual_pointer_v1.c

```diff
@@ -0,0 +1,327 @@
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_motion_absolute event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.x = (double)x / x_extent,
+		.y = (double)y / y_extent,
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.motion_absolute, &event);
+}
+
+static void virtual_pointer_button(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t button,
+		uint32_t state) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	struct wlr_event_pointer_button event = {
+		.device = wlr_dev,
+		.time_msec = time,
+		.button = button,
+		.state = state ? WLR_BUTTON_PRESSED : WLR_BUTTON_RELEASED
+	};
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.button, &event);
+}
+
+static void virtual_pointer_axis(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t source,
+		wl_fixed_t value) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	pointer->axis_event.device = wlr_dev;
+	pointer->axis_event.time_msec = time;
+	pointer->axis_event.source = source;
+	pointer->axis_event.delta = wl_fixed_to_double(value);
+}
+
+static void virtual_pointer_frame(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+
+	if (pointer->axis_event.device) {
+		/* Deliver pending axis event */
+		wlr_signal_emit_safe(&wlr_dev->pointer->events.axis,
+				&pointer->axis_event);
+		memset(&pointer->axis_event, 0, sizeof(pointer->axis_event));
+	}
+
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.frame, wlr_dev->pointer);
+}
+
+static void virtual_pointer_axis_source(struct wl_client *client,
+		struct wl_resource *resource, uint32_t source) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	pointer->axis_event.device = wlr_dev;
+	pointer->axis_event.source = source;
+}
+
+static void virtual_pointer_axis_stop(struct wl_client *client,
+		struct wl_resource *resource, uint32_t time, uint32_t source) {
+	struct wlr_virtual_pointer_v1 *pointer =
+		virtual_pointer_from_resource(resource);
+	if (pointer == NULL) {
+		return;
+	}
+	struct wlr_input_device *wlr_dev = &pointer->input_device;
+	pointer->axis_event.device = wlr_dev;
+	pointer->axis_event.time_msec = time;
+	pointer->axis_event.source = source;
+	wlr_signal_emit_safe(&wlr_dev->pointer->events.axis, &pointer->axis_event);
```

Comment:
I wonder if the protocol should be stricter. It currently allows for things which don't make sense like:

```
axis_stop(time=42, axis=vertical)
axis(time=42, axis=vertical, value=69)
frame()
```

Maybe we should only allow for one axis/axis_stop event per axis (vertical/horizontal) in each frame.